### ISSUE TITLE
~ does not decode as nil for pointer types

### ISSUE THREAD

**Issue by ghost** [go-yaml/yaml#265]
For example, `Value1` is decoded as an empty string when "~" is used instead of "null":

```go
type T1 struct {
	Value1 *string
}

func main() {
	testinfo := map[string]string{
		"canonical": `value1: ~`,
		"english":   `value1: null`,
	}
	for testname, data := range testinfo {
		var foo T1
		if err := yaml.Unmarshal([]byte(data), &foo); err != nil {
			panic(err)
		}
		fmt.Println("==", testname, "==")
		fmt.Println("input: ", data)
		fmt.Printf("output: %+v\n\n", foo)
	}
}
```

A simple fix [decode.go](https://github.com/go-yaml/yaml/blob/v2/decode.go#L253) works:

```go
	if n.tag == yaml_NULL_TAG || n.kind == scalarNode && n.tag == "" && (n.value == "null" || n.value == "~" || n.value == "" && n.implicit) {
```

All current tests pass, including the one above. I believe #109 is related as this code fixes it as well.  I guess someone forgot about fixing this (or it wasn't a priority, which is understandable.)

**Comment by fschuetz04** [comment]
It would be best to open a pull request for that, maybe the maintainers will merge it.

**Comment by rogpeppe** [comment]
Closed by #289
